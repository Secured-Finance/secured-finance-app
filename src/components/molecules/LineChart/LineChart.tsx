import {
    CategoryScale,
    ChartData,
    Chart as ChartJS,
    ChartOptions,
    LineElement,
    LinearScale,
    PointElement,
    Scriptable,
    ScriptableContext,
    Title,
    Tooltip,
} from 'chart.js';
import ChartTooltip from 'chart.js/auto';
import { useCallback, useEffect, useRef } from 'react';
import { ChartProps, Line } from 'react-chartjs-2';
import { Spinner } from 'src/components/atoms';
import {
    crossHairPlugin,
    options as customOptions,
    defaultDatasets,
} from 'src/components/molecules/LineChart/constants';
import { MaturityListItem } from 'src/components/organisms';
import { Maturity } from 'src/utils/entities';

ChartJS.register(
    LinearScale,
    PointElement,
    LineElement,
    Title,
    CategoryScale,
    Tooltip
);

const triggerHover = (chart: ChartJS<'line'>, index: number) => {
    chart.setActiveElements([
        {
            datasetIndex: 0,
            index: index,
        },
    ]);
    chart.update();
};

const triggerTooltip = (chart: ChartJS<'line'>, index: number) => {
    const tooltip = chart.tooltip as unknown as ChartTooltip<'line'>;
    if (tooltip) {
        tooltip.setActiveElements([
            {
                datasetIndex: 0,
                index: index,
            },
        ]);
        tooltip.update();
    }
    chart.update();
};

export type LineChartProps = {
    style?: React.CSSProperties;
    data: ChartData<'line'>;
    maturityList: MaturityListItem[];
    maturity: Maturity;
    handleChartClick: (index: number) => void;
} & ChartProps;

export const LineChart = ({
    data = { datasets: [], labels: [] },
    options = customOptions,
    style,
    maturityList,
    maturity,
    handleChartClick,
}: LineChartProps) => {
    const chartRef = useRef<ChartJS<'line'>>(null);

    if (data.datasets.length > 0 && data.datasets[0].data.length > 0) {
        const dataArray = data.datasets[0].data as number[];

        if (dataArray.every(item => item === 0)) {
            options = {
                ...options,
                scales: {
                    ...options.scales,
                    y: {
                        display: false,
                        beginAtZero: true,
                    },
                },
            };
        }
    }

    const refinedDatasets = data.datasets.map(set => {
        if (defaultDatasets) {
            return {
                borderCapStyle: 'round' as Scriptable<
                    CanvasLineCap,
                    ScriptableContext<'line'>
                >,
                ...defaultDatasets,
                ...set,
            };
        }
        return {
            ...set,
        };
    });

    const refinedData = {
        ...data,
        datasets: refinedDatasets,
    };

    const handleClick = (event: React.MouseEvent<HTMLCanvasElement>) => {
        if (!chartRef.current) return;

        const element = chartRef.current.getElementsAtEventForMode(
            event as unknown as Event,
            'nearest',
            { axis: 'x', intersect: false, includeInvisible: true },
            true
        );
        if (element && element[0]) {
            const { index } = element[0];
            const label = data.labels?.[index];
            const selectedMaturityIndex = maturityList.findIndex(
                element => element.label === label
            );
            if (selectedMaturityIndex >= 0) {
                handleChartClick(selectedMaturityIndex);
            }
        }
    };

    const onMouseOut = useCallback(() => {
        if (!chartRef.current) return;

        const numberOfElements = chartRef.current.data.datasets[0].data.length;
        let index = maturityList.findIndex(
            element => element.maturity === maturity.toNumber()
        );

        if (numberOfElements) {
            index = index > 0 ? index : 0;
            triggerHover(chartRef.current, index);
            triggerTooltip(chartRef.current, index);
        }
    }, [maturityList, maturity]);

    useEffect(() => {
        onMouseOut();
    }, [onMouseOut]);

    return (
        <>
            {refinedData.datasets.length > 0 &&
            refinedData.datasets[0].data.length > 0 ? (
                <Line
                    style={style}
                    data={refinedData}
                    options={options as ChartOptions<'line'>}
                    ref={chartRef}
                    onClick={handleClick}
                    data-chromatic='ignore'
                    plugins={[crossHairPlugin]}
                    onMouseOut={onMouseOut}
                />
            ) : (
                <div className='flex h-full w-full items-center justify-center'>
                    <Spinner />
                </div>
            )}
        </>
    );
};
