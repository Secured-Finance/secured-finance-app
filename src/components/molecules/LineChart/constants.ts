import {
    ActiveElement,
    Chart,
    ChartData,
    ChartEvent,
    ChartOptions,
    ChartTypeRegistry,
    ScriptableContext,
    TooltipItem,
} from 'chart.js';
import { Rate } from 'src/utils';

export const defaultDatasets = {
    borderWidth: 3,
    radius: 0,
    lineTension: 0.4,
    pointRadius: 0.01,
};

const refineArray = (array: Array<Rate>) => {
    return array.map(r => r.toNormalizedNumber());
};

export const getData = (
    rates: Rate[],
    label: string,
    labels: string[]
): ChartData<'line'> => {
    return {
        labels: labels,
        datasets: [
            {
                label: label,
                data: refineArray(rates),
            },
        ],
    };
};

export const crossHairPlugin = {
    id: 'cross-hair',
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    beforeDraw: (chart: any) => {
        if (
            chart.tooltip &&
            chart.tooltip._active &&
            chart.tooltip._active.length
        ) {
            const activePoint = chart.tooltip._active[0];
            const ctx = chart.ctx;
            const x = activePoint.element.x;
            const y = activePoint.element.y;
            const bottomY = chart.scales.y.bottom;
            ctx.save();
            ctx.beginPath();
            ctx.moveTo(x, y);
            ctx.lineTo(x, bottomY + 15);
            ctx.lineWidth = 1;
            ctx.setLineDash([8, 8]);
            ctx.strokeStyle = 'rgba(252, 252, 253, 1)';
            ctx.stroke();
            ctx.restore();
        }
    },
};

export const crossHairMultiPlugin = {
    id: 'cross-hair-multi',
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    beforeDraw: (chart: any) => {
        if (
            chart.tooltip &&
            chart.tooltip._active &&
            chart.tooltip._active.length
        ) {
            const activePoint = chart.tooltip._active[0];
            const ctx = chart.ctx;
            const x = activePoint.element.x;
            const topY = chart.scales.y.top;
            const bottomY = chart.scales.y.bottom;
            ctx.save();
            ctx.beginPath();
            ctx.moveTo(x, topY);
            ctx.lineTo(x, bottomY);
            ctx.lineWidth = 1;
            ctx.setLineDash([8, 8]);
            ctx.strokeStyle = '#B9BDEA';
            ctx.stroke();
            ctx.restore();
        }
    },
};

export const getCurveGradient = (context: ScriptableContext<'line'>) => {
    const ctx = context.chart.ctx;
    const yieldCurveGradient = ctx.createLinearGradient(
        0,
        0,
        ctx.canvas.offsetWidth,
        0
    );
    yieldCurveGradient.addColorStop(0, 'rgba(255, 89, 248, 0)');
    yieldCurveGradient.addColorStop(0.2, 'rgba(174, 114, 255, 1)');
    yieldCurveGradient.addColorStop(0.5, 'rgba(144, 233, 237, 1)');
    yieldCurveGradient.addColorStop(0.78, 'rgba(92, 209, 103, 1)');
    yieldCurveGradient.addColorStop(1, 'rgba(255, 238, 0, 0)');
    return yieldCurveGradient;
};

export const options: ChartOptions<'line'> = {
    responsive: true,
    maintainAspectRatio: false,
    interaction: {
        mode: 'index',
        intersect: false,
    },
    layout: {
        padding: {
            top: 50,
        },
    },
    elements: {
        point: {
            hoverRadius: 6,
            hoverBorderColor: '#FFFFFF',
            hoverBorderWidth: 2,
            backgroundColor: '#5162FF',
        },
    },
    scales: {
        y: {
            display: false,
        },

        x: {
            beginAtZero: true,
            ticks: {
                color: 'rgba(255, 255, 255, 0.6)',
                font: {
                    lineHeight: 2.0,
                },
                padding: 10,
            },
            grid: {
                display: false,
                borderColor: 'transparent',
            },
        },
    },
    hover: {
        intersect: false,
    },
    onHover: (
        _event: ChartEvent,
        _elements: ActiveElement[],
        chart: Chart<'line'>
    ) => {
        chart.canvas.style.cursor = 'pointer';
    },
    plugins: {
        tooltip: {
            yAlign: 'bottom',
            caretPadding: 16,
            backgroundColor: 'rgba(47, 50, 65, 1)',
            borderWidth: 1,
            borderColor: 'rgba(52, 56, 76, 1)',
            displayColors: false,
            cornerRadius: 10,
            padding: 8,
            callbacks: {
                label: (item: TooltipItem<keyof ChartTypeRegistry>) => {
                    let content = '';
                    content += item.formattedValue + '%';
                    return content;
                },
                title: () => {
                    return '';
                },
            },
        },
    },
    events: ['click', 'mousemove'],
};
