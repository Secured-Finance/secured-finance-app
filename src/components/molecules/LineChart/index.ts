export * from './constants';
export * from './LineChart';
