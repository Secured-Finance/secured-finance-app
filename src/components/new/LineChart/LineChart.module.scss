@import '../../../theme/styleduide';

.container {
    position: relative;
}

.chartView {
    margin-top: $paddingS;
}

.title {
    color: $lightSilver;
    font-size: $tabSize;
}

.legendMarker {
    width: 20px;
    height: 20px;
}

.legend {
    position: absolute;
    right: 24px;
    display: inline-block;
    bottom: 40px;
    color: $lightGrey;
    font-size: $captionSize;

    > ul {
        display: flex;
        list-style-type: none;

        :not(:first-child) {
            margin-left: $paddingSM;
        }

        > li {
            display: flex;

            > span {
                display: inline-flex;
                width: 5px;
                height: 5px;
                border-radius: 50%;
                align-self: center;
                margin-right: $paddingXS;
            }
        }
    }
}

@media (max-width: 1024px) {
    .container {
        width: 520px;
    }
}

@media (max-width: 1280px) {
    .container {
        width: 600px;
    }
}
