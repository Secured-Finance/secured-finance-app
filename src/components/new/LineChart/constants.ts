const canvas: HTMLCanvasElement = document.createElement('canvas');
let ctx: CanvasRenderingContext2D = canvas.getContext('2d');

const purpleGradient = ctx.createLinearGradient(0, 0, 0, 290);
purpleGradient.addColorStop(0, 'rgba(91, 80, 222, 0.5)');
purpleGradient.addColorStop(1, 'rgba(15, 26, 34, 0.1)');

const greenGradient = ctx.createLinearGradient(0, 0, 0, 290);
greenGradient.addColorStop(0, 'rgba(147, 194, 87, 1)');
greenGradient.addColorStop(1, 'rgba(15, 26, 34, 0.1)');

const redGradient = ctx.createLinearGradient(0, 0, 0, 290);
redGradient.addColorStop(0, 'rgba(222, 95, 66, 1)');
redGradient.addColorStop(1, 'rgba(15, 26, 34, 0.1)');

const yAxisGradient = ctx.createLinearGradient(0, 100, 600, 100);
yAxisGradient.addColorStop(0, 'rgba(53, 54, 61, 0.3)');
yAxisGradient.addColorStop(0.9, 'rgba(53, 54, 61, 0.1)');
yAxisGradient.addColorStop(1, 'rgba(53, 54, 61, 0)');

const xAxisGradient = ctx.createLinearGradient(0, 100, 600, 100);
xAxisGradient.addColorStop(0, 'rgba(53, 54, 52, 0.3)');
xAxisGradient.addColorStop(0.05, 'rgba(53, 54, 61, 0)');

const xAxisZeroLineGradient = ctx.createLinearGradient(300, 200, 300, 0);
xAxisZeroLineGradient.addColorStop(0.6, 'rgba(53, 54, 61, 0.3)');
xAxisZeroLineGradient.addColorStop(1, 'rgba(53, 54, 61, 0)');

export const commonDataset = {
    fill: true,
    pointHitRadius: 100,
    borderWidth: 2,
    opacity: 1,
    hidden: false,
    radius: 0,
    hoverRadius: 3,
};

export const defaultDatasets = [
    {
        ...commonDataset,
        backgroundColor: purpleGradient,
        borderColor: '#5b50de',
    },
    {
        ...commonDataset,
        backgroundColor: redGradient,
        borderColor: '#de5f42',
    },
    {
        ...commonDataset,
        backgroundColor: greenGradient,
        borderColor: '#93c257',
    },
];

const getYAxisLinesColor = () => {
    // hack for hiding the last yAxis line for charts of 20 ticks max
    const colors: Array<string | CanvasGradient> = ['transparent'];
    for (let i = 0; i < 20; i++) {
        colors.push(yAxisGradient);
    }

    return colors;
};

export const options = {
    responsive: true,
    scales: {
        yAxes: [
            {
                color: yAxisGradient,
                ticks: {
                    color: '#5f616a',
                    fontSize: 13,
                    padding: 8,
                    callback: (value: number) => value + '%',
                },
                gridLines: {
                    color: getYAxisLinesColor(),
                    drawBorder: false,
                    zeroLineColor: yAxisGradient,
                    drawTicks: false,
                },
            },
        ],
        xAxes: [
            {
                ticks: {
                    color: '#5f616a',
                    fontSize: 13,
                    padding: 8,
                },
                gridLines: {
                    color: xAxisGradient, // hides xAxis grid except for zero line
                    drawBorder: false,
                    zeroLineColor: xAxisZeroLineGradient,
                    drawTicks: false,
                },
            },
        ],
    },
    tooltips: {
        filter: (item: any) => {
            return item.value > 0;
        },
        callbacks: {
            label: ({ value, datasetIndex, yLabel }: any, data: any) => {
                if (value > 0) {
                    const label = data.datasets[datasetIndex].label || '';
                    let content = '';

                    if (data.datasets.length > 1) {
                        content += label;
                    }

                    content += ' ' + yLabel + ' %';
                    return content;
                }
                return false;
            },
        },
    },
    legend: {
        display: false,
        position: 'right',
    },
    legendCallback: (chart: any) => {
        const ul = document.createElement('ul');
        chart.data?.datasets?.forEach((dataset: any) => {
            const borderColor = dataset.borderColor;
            ul.innerHTML += `
                <li>
                   <span class="legendMarker" style="background-color: ${borderColor}"></span>
                    ${dataset.label}
                 </li>
              `;
        });
        return ul.outerHTML;
    },
};
