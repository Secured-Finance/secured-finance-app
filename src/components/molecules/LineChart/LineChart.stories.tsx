import { ComponentMeta, ComponentStory } from '@storybook/react';
import { options } from '../../molecules/LineChart/constants';
import { LineChart } from './';

export default {
    title: 'Molecules/LineChart',
    component: LineChart,
    args: {
        data: {
            labels: [
                'SEP22',
                'DEC22',
                'MAR23',
                'JUN23',
                'SEP23',
                'DEC23',
                'MAR24',
                'JUN24',
            ],
            datasets: [
                {
                    label: 'Borrow',
                    data: [50, 34, 27.37, 22.29, 18, 14.22, 10.81, 8.2],
                },
            ],
        },
        options,
    },
} as ComponentMeta<typeof LineChart>;

const Template: ComponentStory<typeof LineChart> = args => (
    <div style={{ width: 500, height: 350 }}>
        <LineChart {...args} />
    </div>
);

export const Default = Template.bind({});
