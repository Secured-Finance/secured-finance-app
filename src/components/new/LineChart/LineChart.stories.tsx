import { Meta, Story } from '@storybook/react';
import { LineChart, LineChartProps } from './';
import { options } from './constants';

export default {
    title: 'Components/LineChart',
    component: LineChart,
} as Meta;

const Template: Story<LineChartProps> = args => (
    <div style={{ width: 700 }}>
        <LineChart {...args} />
    </div>
);
export const Default = Template.bind({});
Default.args = {
    title: 'Yield Curve',
    data: {
        labels: ['0', '3m', '6m', '1y', '2y', '3y', '5y'],
        datasets: [
            {
                label: 'Borrow',
                data: [0, 7, 8, 9, 10, 13, 15],
            },
            // {
            //     label: 'Lend',
            //     data: [0, 9.2, 10.2, 11.2, 12.2, 14, 15.4],
            // },
            // {
            //     label: 'Spread',
            //     data: [0, 8.1, 9.1, 10.1, 11.1, 13.5, 15.2],
            // },
        ],
    },
    options,
    style: {},
    showLegend: true,
};
