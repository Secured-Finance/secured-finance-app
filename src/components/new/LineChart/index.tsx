import React, { useEffect, useRef } from 'react';
import cm from './LineChart.module.scss';
import { Line, ChartComponentProps } from 'react-chartjs-2';
import { defaultDatasets, commonDataset } from './constants';
import { options } from './constants';

export interface ILineChart extends ChartComponentProps {
    title?: string | JSX.Element;
    style?: React.CSSProperties;
    showLegend?: boolean;
}

export const LineChart: React.FC<ILineChart> = ({
    data,
    options,
    title,
    style,
    showLegend,
}) => {
    const refinedDatasets = data.datasets.map((set: any, i: number) => {
        if (defaultDatasets[i]) {
            return {
                ...defaultDatasets[i],
                ...set,
            };
        }
        return {
            ...commonDataset,
            ...set,
        };
    });
    const refinedData = {
        ...data,
        datasets: refinedDatasets,
    };

    const chartRef = useRef(null);
    useEffect(() => {
        if (showLegend && data.datasets.length) {
            document.getElementById('legend').innerHTML =
                chartRef.current.chartInstance.generateLegend();
        }
    }, [showLegend, data.datasets.length]);

    return (
        <div className={cm.container} style={style}>
            {title && <span className={cm.title}>{title}</span>}
            <div className={cm.chartView}>
                <Line data={refinedData} options={options} ref={chartRef} />
                {showLegend && <div id={'legend'} className={cm.legend} />}
            </div>
        </div>
    );
};

LineChart.defaultProps = {
    data: {
        datasets: [],
    },
    options,
};
