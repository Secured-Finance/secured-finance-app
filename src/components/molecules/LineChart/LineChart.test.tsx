import { composeStories } from '@storybook/testing-react';
import { queries, render, RenderResult, waitFor } from 'src/test-utils.js';
import * as stories from './LineChart.stories';

const { Default } = composeStories(stories);

describe('LineChart Component', () => {
    it.skip('should render a LineChart', async () => {
        let ag: RenderResult<typeof queries, HTMLElement>;
        await waitFor(() => {
            ag = render(<Default />);
        }).then(() => {
            expect(
                ag.container
                    .querySelector('canvas')
                    ?.getContext('2d')
                    ?.__getEvents()
            ).toMatchSnapshot();
        });
    });
});
