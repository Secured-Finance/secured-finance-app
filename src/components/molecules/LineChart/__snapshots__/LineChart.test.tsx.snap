// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart Component should render a LineChart 1`] = `
[
  {
    "props": {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "value": "left",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  {
    "props": {
      "value": "middle",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  {
    "props": {
      "text": "Borrow",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "0",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "50",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "value": "left",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  {
    "props": {
      "value": "middle",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  {
    "props": {
      "text": "Borrow",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "SEP22",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "DEC22",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "MAR23",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "JUN23",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "SEP23",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "DEC23",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "MAR24",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "JUN24",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px "Helvetica Neue", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
]
`;
